### ISSUE TITLE
Dynamic resources in Python don't work outside of __main__

### ISSUE THREAD

**Issue by benesch** [pulumi/pulumi#7453]
Instantiating a dynamic resource in Python outside of the `__main__` module fails. This was filed as #4100 but that issue was incorrectly closed, in my opinion! The repro there is totally legit and I am copy/pasting it below.

## Steps to reproduce

Run `pulumi up` on this program:

```py
# __main__.py
import custom_resource

# custom_resource.py
import pulumi
import pulumi.dynamic


class CustomDynamicResourceProvider(
    pulumi.dynamic.ResourceProvider,
):
    pass


pulumi.dynamic.Resource(
    provider=CustomDynamicResourceProvider(),
    name='my-resource',
    props={},
)
```

It will incorrectly fail:

```
Previewing update (dev)

View Live: https://app.pulumi.com/benesch/pulumi-4100/dev/previews/b9c38f39-280d-40d2-9bed-e6c5ccd7636c

     Type                               Name             Plan       Info
 +   pulumi:pulumi:Stack                pulumi-4100-dev  create     
     └─ pulumi-python:dynamic:Resource  my-resource                 1 error
 
Diagnostics:
  pulumi-python:dynamic:Resource (my-resource):
    error: Exception calling application: Program run without the Pulumi engine available; re-run using the `pulumi` CLI
```

I've put this together in a Git repository, if that's easier: https://github.com/benesch/pulumi-4100

As far as I can tell, all the dynamic resource examples in the pulumi/examples repository do so from `__main__` so this bug is never observed.

cc @leezen @ilaif

**Comment by komalali** [comment]
Looks like this is caused by `register_resource` for the dynamic resource being called before the monitor address has been established, so when `register_resource` calls [get_monitor](https://github.com/pulumi/pulumi/blob/eb32039013277d8bce69fe284e93fbdbfcab9964/sdk/python/lib/pulumi/runtime/resource.py#L420) it fails with that error. 

I’m surprised that this is possible, and it’s not immediately clear why the program is parsed and imports are loaded before `runpy` is called on `__main__` (where the instantiated dynamic resource is actually imported).

**Comment by komalali** [comment]
Some more digging, it looks like when an instantiated dynamic resource is imported into `__main__`, the code in https://github.com/pulumi/pulumi/blob/480173a57f7360c4f261c1b28efbb02cd4381bb5/sdk/python/cmd/pulumi-language-python-exec (the entrypoint) never runs. 

Even if it did though, this wouldn't work when using inline programs in automation API since the module is loaded prior to execution of the pulumi program (which requires the engine).

I'd love to understand what necessitates instantiating the dynamic resource in another module.

**Comment by benesch** [comment]
> I'd love to understand what necessitates instantiating the dynamic resource in another module.

Our Pulumi app is many thousands of lines of code, so we’ve split it up into a dozen or so modules to keep it maintainable. The point at which we need to instantiate the resource is “between” modules: it uses some of the modules but is depended upon by others. So it would be a rather painful refactor to instantiate this resource in __main__.

Is there some other recommended pattern for structuring Pulumi programs? Very happy to take advice on that front.

**Comment by komalali** [comment]
Hmm... okay I just noticed something else interesting... I don't see the same error if the definition of the dynamic resource is in a different module than the instantiation.

Concretely, the following results in the same error you mentioned:
```python
# __main__.py
import component

# component.py
from pulumi.dynamic import Resource, ResourceProvider, CreateResult


class CustomDynamicResourceProvider(ResourceProvider):
    def create(self, props):
        return CreateResult("1", {"val": "blah"})


class CustomDynamicResource(Resource):
    val: str

    def __init__(self, name, opts=None):
        super().__init__(CustomDynamicResourceProvider(), name, {"val": ""}, opts)


CustomDynamicResource("foo")
```

However, the following succeeds:
```python
# __main__.py
import component

# component.py
from another import CustomDynamicResource

CustomDynamicResource("foo")

# another.py
from pulumi.dynamic import Resource, ResourceProvider, CreateResult


class CustomDynamicResourceProvider(ResourceProvider):
    def create(self, props):
        return CreateResult("1", {"val": "blah"})


class CustomDynamicResource(Resource):
    val: str

    def __init__(self, name, opts=None):
        super().__init__(CustomDynamicResourceProvider(), name, {"val": ""}, opts)
```

I'm curious if that workaround would work for you.

**Comment by benesch** [comment]
Ooh, that _is_ interesting. That could do the trick!

**Comment by jancespivo** [comment]
Hi, I also struggle with the similar issue. We also use automation in a dynamic resource provider and also with Python. Sometime some of resources fail with `Program run without the Pulumi engine available; re-run using the pulumi CLI`. It is enough to repeat `pulumi up` and the failed resources are updated successfully.

**Comment by joeduffy** [comment]
I was curious about this one. I stumbled upon https://github.com/uqfoundation/dill/issues/56, which talks about some of the ways in which dill serializes/deserializes classes differently for `__main__` versus modules. What I see happening is that when dill tries to deserialize the class, via [this call to `dill.loads`](https://github.com/pulumi/pulumi/blob/7b7f03ea55eb323d84b2be8e4f7b11525e489a23/sdk/python/lib/pulumi/dynamic/__main__.py#L38), it tries to run the module initialization all over again which results in a call to `register_resource` from within the dynamic provider process itself, which should never happen (all we want to do is serve the gRPC server so calls to Check/Diff/Create/etc can be called). Apparently this doesn't happen the same way for `__main__`.

If we fake out dill by changing the class's module to `__main__` during serialization, things "seem to work" (I have no clue what other ramifications this might have). For example, this patch works for any class:

```diff
diff --git a/sdk/python/lib/pulumi/dynamic/dynamic.py b/sdk/python/lib/pulumi/dynamic/dynamic.py
index a58e17604..5b90f670f 100644
--- a/sdk/python/lib/pulumi/dynamic/dynamic.py
+++ b/sdk/python/lib/pulumi/dynamic/dynamic.py
@@ -222,10 +222,16 @@ def serialize_provider(provider: ResourceProvider) -> str:
     pickle.Pickler.save_dict = save_dict_sorted

     # Use dill to recursively pickle the provider and store base64 encoded form
+    pty = type(provider)
+    oldptymod = pty.__module__
     try:
+        pty.__module__ = '__main__'
         byts = dill.dumps(provider, protocol=pickle.DEFAULT_PROTOCOL, recurse=True)
         return base64.b64encode(byts).decode('utf-8')
     finally:
+        pty.__module__ = oldptymod
         # Restore the original pickler
         pickle.Pickler = old_pickler
```

With these changes, @benesch's original example "works". You can also do it by hand:

```python
# __main__.py
import custom_resource

# custom_resource.py
import pulumi
import pulumi.dynamic

class CustomDynamicResourceProvider(
    pulumi.dynamic.ResourceProvider,
):
    pass

CustomDynamicResourceProvider.__module__ = '__main__'
pulumi.dynamic.Resource(
    provider=CustomDynamicResourceProvider(),
    name='my-resource',
    props={},
)
CustomDynamicResourceProvider.__module__ = 'custom_resource'
```

Again, this seems rather dubious to do, but I wanted to drop a note in case it helps with further investigation.

**Comment by benesch** [comment]
Wow, thanks for digging so deep, @joeduffy! I played around with your discovery in the `pulumi-4100` repo I linked to above. When setting the `__module__` to `__main__`, this is the disassembly of the pickled provider class:

```
        0: \x80 PROTO      4
        2: \x95 FRAME      381
       11: \x8c SHORT_BINUNICODE 'dill._dill'
       23: \x94 MEMOIZE    (as 0)
       24: \x8c SHORT_BINUNICODE '_create_type'
       38: \x94 MEMOIZE    (as 1)
       39: \x93 STACK_GLOBAL
       40: \x94 MEMOIZE    (as 2)
       41: (    MARK
       42: h        BINGET     0
       44: \x8c     SHORT_BINUNICODE '_load_type'
       56: \x94     MEMOIZE    (as 3)
      ... snip ...
      384: t        TUPLE      (MARK at 41)
      385: \x94 MEMOIZE    (as 36)
      386: R    REDUCE
      387: \x94 MEMOIZE    (as 37)
      388: )    EMPTY_TUPLE
      389: \x81 NEWOBJ
      390: \x94 MEMOIZE    (as 38)
      391: .    STOP
    highest protocol among opcodes = 4
```

Whereas if you leave `__module__` alone, you get this instead:

```
        0: \x80 PROTO      4
        2: \x95 FRAME      56
       11: \x8c SHORT_BINUNICODE 'custom_resource'
       28: \x94 MEMOIZE    (as 0)
       29: \x8c SHORT_BINUNICODE 'CustomDynamicResourceProvider'
       60: \x94 MEMOIZE    (as 1)
       61: \x93 STACK_GLOBAL
       62: \x94 MEMOIZE    (as 2)
       63: )    EMPTY_TUPLE
       64: \x81 NEWOBJ
       65: \x94 MEMOIZE    (as 3)
       66: .    STOP
    highest protocol among opcodes = 4
```

The tl;dr seems to be that dill serializes classes in main by *value*, while it serializes classes in modules by *reference*. And so it all makes perfect sense: if the serialized representation is `custom_resource.CustomProvider`, dill of course has no choice but to run `from custom_resource import CustomProvider` when unpicking. Hrm.

**Comment by benesch** [comment]
`cloudpickle` allows you to [control what modules pickle by value](https://github.com/cloudpipe/cloudpickle#overriding-pickles-serialization-mechanism-for-importable-constructs), so perhaps the solution is as simple as switching to `cloudpickle`. Let me give that a whirl.

**Comment by benesch** [comment]
So, I tried this diff:

```diff
diff --git a/sdk/python/Pipfile b/sdk/python/Pipfile
index 2a17fd2e6..52391ddca 100644
--- a/sdk/python/Pipfile
+++ b/sdk/python/Pipfile
@@ -7,7 +7,7 @@ name = "pypi"
 # Keep this list in sync with setup.py
 protobuf = ">=3.6.0"
 grpcio = ">=1.33.2"
-dill = ">=0.3.0"
+cloudpickle = { git = "https://github.com/cloudpipe/cloudpickle.git", ref = "343da119685f622da2d1658ef7b3e2516a01817f" }
 six = ">=1.12.0"
 semver = ">=2.8.1"
 pyyaml = ">=5.3.1"
diff --git a/sdk/python/lib/pulumi/dynamic/__main__.py b/sdk/python/lib/pulumi/dynamic/__main__.py
index 6fbac4b92..8d19430c8 100644
--- a/sdk/python/lib/pulumi/dynamic/__main__.py
+++ b/sdk/python/lib/pulumi/dynamic/__main__.py
@@ -15,10 +15,10 @@
 import asyncio
 import base64
 from concurrent import futures
+import pickle
 import sys
 import time
 
-import dill
 import grpc
 from google.protobuf import empty_pb2
 from pulumi.runtime import proto, rpc
@@ -35,7 +35,7 @@ _GRPC_CHANNEL_OPTIONS = [('grpc.max_receive_message_length', _MAX_RPC_MESSAGE_SI
 
 def get_provider(props) -> ResourceProvider:
     byts = base64.b64decode(props[PROVIDER_KEY])
-    return dill.loads(byts)
+    return pickle.loads(byts)
 
 class DynamicResourceProviderServicer(ResourceProviderServicer):
     def CheckConfig(self, request, context):
diff --git a/sdk/python/lib/pulumi/dynamic/dynamic.py b/sdk/python/lib/pulumi/dynamic/dynamic.py
index a58e17604..42f48b2b5 100644
--- a/sdk/python/lib/pulumi/dynamic/dynamic.py
+++ b/sdk/python/lib/pulumi/dynamic/dynamic.py
@@ -15,9 +15,10 @@
 import asyncio
 import base64
 import pickle
+import sys
 from typing import Any, Optional, List, TYPE_CHECKING, no_type_check, cast
 
-import dill
+import cloudpickle
 from .. import CustomResource, ResourceOptions
 
 if TYPE_CHECKING:
@@ -221,9 +222,11 @@ def serialize_provider(provider: ResourceProvider) -> str:
         self._batch_setitems(sorted(obj.items())) # pylint: disable=protected-access
     pickle.Pickler.save_dict = save_dict_sorted
 
-    # Use dill to recursively pickle the provider and store base64 encoded form
+    # Use cloudpickle to recursively pickle the provider and store base64 encoded
+    # form
     try:
-        byts = dill.dumps(provider, protocol=pickle.DEFAULT_PROTOCOL, recurse=True)
+        cloudpickle.register_pickle_by_value(sys.modules[provider.__module__])
+        byts = cloudpickle.dumps(provider, protocol=pickle.DEFAULT_PROTOCOL)
         return base64.b64encode(byts).decode('utf-8')
     finally:
         # Restore the original pickler
diff --git a/sdk/python/lib/setup.py b/sdk/python/lib/setup.py
index 0f67cbcb7..c49d02e59 100644
--- a/sdk/python/lib/setup.py
+++ b/sdk/python/lib/setup.py
@@ -38,7 +38,7 @@ setup(name='pulumi',
       # Keep this list in sync with Pipfile
       install_requires=[
           'protobuf>=3.6.0',
-          'dill>=0.3.0',
+          'cloudpickle @ git+https://github.com/cloudpipe/cloudpickle.git#343da119685f622da2d1658ef7b3e2516a01817f',
           'grpcio>=1.33.2',
           'six>=1.12.0',
           'semver>=2.8.1',
```

This actually fixes this bug! But it introduces another, as cloudpickle doesn't promise a deterministic serialization, and so every `pulumi up` generates a spurious provider diff.

I opened a feature request against dill about the ability to force force by-value serialization: https://github.com/uqfoundation/dill/issues/424. But @joeduffy I feel like your solution could be perfectly reasonable in the meantime.

**Comment by carlmontanari** [comment]
I'm in over my head on this as I just started playing with Pulumi this week and am very much unclear on all the magic happening on the backend of things... however... @komalali's fix above didn't work for my slightly different setup, but I stumbled into another workaround. 

My project directory looks like a "normal" python library (perhaps this is a bad idea to start with 😁), so something like this:

```
.
├── Pulumi.my_super_cool_project.yaml
├── Pulumi.yaml
├── __main__.py
├── requirements.txt
├── setup.cfg
├── setup.py
└── my_super_cool_project
    ├── __init__.py
    ├── containers
    │   ├── __init__.py
    │   └── registry.py
    ├── db
    │   ├── __init__.py
    │   ├── postgres.py
    │   └── sql.py
    ├── network
    │   ├── __init__.py
    │   └── vpc.py
    └── storage
        ├── __init__.py
        └── buckets.py
```

The `my_super_cool_project/db/postgres.py` module contains a custom resource I'm playing around with. My original `__main__` was just this:

```
from containers import registry
from db import sql
from network import vpc
from storage import buckets
```

The `__init__` in `my_super_cool_project` just imported all this stuff and stuffed it in `__all__` so I could be lazy about importing it in `__main__`.

Prior to adding the custom resource this worked just fine. After adding that custom resource I got the same `Program run without the Pulumi engine available` others have reported.

Moving the `postgres` file up to `my_super_cool_project/custom_resources/postgres.py` and changing imports in the `sql.py` file where I create the Postgres instance and such didn't have any impact (tried this as it seemed like having things in different module was working for other setups).

Chaining my `__init__` file in `my_super_cool_project` to look like:

```
def run():
    from werkshop_deploy.containers import registry
    from werkshop_deploy.db import sql
    from werkshop_deploy.network import vpc
    from werkshop_deploy.storage import buckets

__all__ = (
    "run",
)
```

And then just importing and executing `run` in the `__main__` does seem to work with my custom resource. Now I just gotta get the custom resource working 🤣 .

Like I said -- way outa my element here but figured this may be useful for somebody!

Carl

**Comment by lukehoban** [comment]
> This (`cloudpickle`) actually fixes this bug! But it introduces another, as cloudpickle doesn't promise a deterministic serialization, and so every pulumi up generates a spurious provider diff.

Yes - I recall when we initially implemented closure serialization support for Python, we initially looked at `cloudpickle` but were blocked on using it by this non-determinism.  Changing to `cloudpickple` would also potentially be a non-trivial breaking change for any existing deployed projects, which would require some very careful handling to switch over to - so I expect this is not currently the path we want to go.

> I opened a feature request against dill about the ability to force force by-value serialization: uqfoundation/dill#424. But @joeduffy I feel like your solution could be perfectly reasonable in the meantime.

Given that dynamic resources do not work outside of `__main__` today, I cannot see any reason why @joeduffy's proposal here would break anything that does already work, and is likely a safe change to make at least until uqfoundation/dill#424 is addressed and provides a way to support pickled classes outside of `__main__` with "correct" `__module__` (though I do not expect there are many/any practical implications of the change of `__module__` here - it is still not fully aligned with what the user intended).

I suggest we attempt to take a patch with the changes @joeduffy suggested for now, while waiting on uqfoundation/dill#424 for a more complete fix.

**Comment by lukehoban** [comment]
We had to revert the previous change here due to the issues it caused in https://github.com/pulumi/pulumi/issues/7795.  Re-opening this as we'll need to come up with a deeper fix, potentially involving addressing this directly in `dill` itself.

**Comment by kingwill101** [comment]
Is there a workaround for this issue? Currently hitting this issue


```python
import binascii
import os

import pulumi
from godaddypy import Client, Account
from pulumi import Input, Output
from pulumi.dynamic import *


def get_client() -> Client:
    config = pulumi.Config()
    api_key = config.require_secret("godaddy_key")
    api_secret = config.require_secret("godaddy_secret")
    godaddyClient = Client()

    def setupClient(self, args):
        account = Account(args[0], api_secret=args[1])
        godaddyClient = Client(account)

    pulumi.Output.all(api_key, api_secret).all(setupClient)

    return godaddyClient


class GodaddyInputs(object):
    domain: Input[str]
    value: Input[str]
    name: Input[str]
    ttl: Input[str]
    ttype: Input[int]

    def __init__(self, domain, value, name, ttl, ttype):
        self.domain = domain
        self.value = value
        self.name = name
        self.ttl = ttl
        self.ttype = ttype


class GoDaddyProvider(ResourceProvider):

    def has_record(self, args):
        domain = args["domain"]
        name = args["name"]
        ttype = args["ttype"]

        records = get_client().get_records(domain)

        found = False

        for record in records:
            if record.get("name") is not None and record.get("type") is not None:
                if record["name"] == name and record["type"] == ttype:
                    found = True
                    break

        return found

    def create(self, args):

        if not self.has_record(args):
            get_client(). \
                add_record(args["domain"],
                           {
                               'data': args["value"],
                               'name': args["name"],
                               'ttl': args["ttl"],
                               'type': args["ttype"]
                           })

        return CreateResult("schema-" + binascii.b2a_hex(os.urandom(16)).decode("utf-8"), outs=args)

    def delete(self, id, args):
        if self.has_record(args):
            get_client().delete_records(args["domain"], args["name"])

    def diff(self, id, old_inputs, new_inputs):

        replaces = []
        if (old_inputs["domain"] != new_inputs["domain"]): replaces.append("domain")
        if (old_inputs["ttype"] != new_inputs["ttype"]): replaces.append("ttype")
        if (old_inputs["value"] != new_inputs["value"]): replaces.append("value")
        if (old_inputs["name"] != new_inputs["name"]): replaces.append("name")

        return DiffResult(
            changes=old_inputs != new_inputs,
            replaces=replaces,
            stables=None,
            delete_before_replace=True)

    def update(self, id, old_inputs, new_inputs):
        return UpdateResult(outs={**new_inputs})


class GoDaddy(Resource):
    domain: Output[str]
    value: Output[str]
    name: Output[str]
    ttl: Output[str]
    ttype: Output[int]

    def __init__(self, resource_name: str, args: GodaddyInputs, opts=None):
        super().__init__(GoDaddyProvider(), resource_name, vars(args), opts)
```

**Comment by cojam-dutchie** [comment]
I just bumped into this myself and it occurred to me that symlinking the provider into the main module might work... And it did! It's a little silly but it's also not the first time I've used symlinks to mount code lol. And I suppose this isn't as bad due to the fact that the symlinked file is also a normal dependency of the project. Anyway, I just tested my theory really quick and it seems workable so I thought I'd share.

The bug...

```
from foo.dynamic_provider import DynamicResource

DynamicResource()
```

> Exception calling application: No module named 'foo'

The fix...

```
ln -s ./venv/lib/foo/dynamic_provider.py dynamic_provider.py
```

```
from dynamic_provider import DynamicResource

DynamicResource()
```
